core/unix/rseq_linux.c

```diff
@@ -468,7 +549,13 @@ rseq_process_elf_sections(module_area_t *ma, bool at_map,
      * relocated and only need the offset if relocations have not yet been applied.
      */
     ssize_t entry_offs = 0;
-    if (at_map || (DYNAMO_OPTION(early_inject) && !dr_api_entry && !dynamo_started))
+    bool adjust_entry_offs =
+        at_map || (DYNAMO_OPTION(early_inject) && !dr_api_entry && !dynamo_started);
+    if (have_glibc_rseq) {
+        /* We want to set entry_offs if glibc has not been initialized yet. */
+        adjust_entry_offs |= (glibc_rseq_offset == 0);
```

Comment:
The module that contains the unrelocated rseq_cs entries is `linux.rseq` (the app binary).

When Glibc's rseq support is enabled, the first rseq system call is made quite early:

```
Breakpoint 1, 0x00007ffff7fdb500 in rseq_register_current_thread (do_rseq=<optimized out>, self=<optimized out>) at ../sysdeps/nptl/dl-tls_init_tp.c:106
106	../sysdeps/nptl/dl-tls_init_tp.c: No such file or directory.
(gdb) bt
#0  0x00007ffff7fdb500 in rseq_register_current_thread (do_rseq=<optimized out>, self=<optimized out>) at ../sysdeps/nptl/dl-tls_init_tp.c:106
#1  __tls_init_tp () at ../sysdeps/nptl/dl-tls_init_tp.c:108
#2  0x00007ffff7fe52f4 in init_tls (naudit=naudit@entry=0) at ./elf/rtld.c:821
#3  0x00007ffff7fe8c02 in dl_main (phdr=<optimized out>, phnum=<optimized out>, user_entry=<optimized out>, auxv=<optimized out>) at ./elf/rtld.c:2091
#4  0x00007ffff7fe3f2f in _dl_sysdep_start (start_argptr=start_argptr@entry=0x7fffffffdcc0, dl_main=dl_main@entry=0x7ffff7fe5db0 <dl_main>) at ../sysdeps/unix/sysv/linux/dl-sysdep.c:140
#5  0x00007ffff7fe5b21 in _dl_start_final (arg=0x7fffffffdcc0) at ./elf/rtld.c:507
#6  _dl_start (arg=0x7fffffffdcc0) at ./elf/rtld.c:596
#7  0x00007ffff7fe4948 in _start () from /lib64/ld-linux-x86-64.so.2

```

At this time, looks like the rseq_cs entries are not relocated yet. Re the condition in DR: since `dynamo_started` is `true`, it doesn't apply the offset automatically.

Reply:
Hmm, this makes sense, but I'm wondering why we haven't hit this general issue before when examining the executable before ld relocates it: there are some ELF examining functions that just use `at_map` IIRC and I wonder if they get this wrong for the executable pre-ld-reloc.

Reply:
#1589 seems related; it made the distinction between `at_map` and `relocated`. It's marked fixed though.

Reply:
I'm going to submit this PR first so we can call the rseq issue closed.